Adding dataset class

jychoi-hpc opened the pull request:
This is a PR for dataset restructuring by merging "loading rawdata" and "loading serialized data" to avoid intermediate "pkl" file saving.

Brief description of added/updated dataset related classes is as follows:
* BaseDataset as a base dataset class for HydraGNN (an abstract class). 
* RawDataset is based on previous RawDataLoader plus load_serialized_data
* AdiosDataset, PickleDataset, SerializedDataset are added
* Adios writer, pickle writer, and serialized writer are added to save data in different formats

Examples updated to see usages:
* lsms example: updated to use LSMSDataset(RawDataset) and SerializedDataset
* ising example: updated to use LSMSDataset and DistDataset

<img width="1389" alt="image" src="https://user-images.githubusercontent.com/3661063/216803453-bbcec958-dd7d-43ef-b2df-3576d45f96c9.png">

pzhanggit commented on examples/ising_model/train_ising.py:
```diff
@@ -201,41 +219,74 @@ def nsplit(a, n):
+            adwriter.save()
+        elif args.format == "pickle":
+            basedir = os.path.join(os.path.dirname(__file__), "dataset", "pickle")
+            SimplePickleWriter(
```
Do we need to pass `comm=comm` here?

jychoi-hpc replied:
Yes, we need. Thank you for catching!!

pzhanggit reviewed (commented):
Thank you @jychoi-hpc  I really like this PR. I have a few comments mostly on cleaning up stuff in this round. I will need another round to test this changes in runs.

allaffa commented on examples/ising_model/train_ising.py:
```diff
@@ -172,12 +191,11 @@ def nsplit(a, n):
         """
```
The function `nsplit` is used so often in so many examples and utilities...
I would suggest that we create a standalone file in `hydragnn.utils` where we define it once

allaffa commented on hydragnn/utils/rawdataset.py:
```diff
@@ -0,0 +1,695 @@
+from abc import ABC, abstractmethod
+
+
+def tensor_divide(x1, x2):
```
Is `tensor_divide` even used anymore anywhere? 
If it is not used anywhere, we may consider wither deleting it once for all or put it inside `hydragnn.utils` in case it may become useful again one day

jychoi-hpc replied:
I find It is used in two places. I will make it as a common function to be reusable.

allaffa commented on hydragnn/utils/rawdataset.py:
```diff
@@ -0,0 +1,695 @@
+
+
+## All-reduce with numpy array
+def comm_reduce(x, op):
```
Should `comm_reduce` also be moved to a separate file inside `hydrogen.utils`?

jychoi-hpc replied:
Great. I find two places used. I will make it as a common function as you suggest.

allaffa commented on hydragnn/utils/rawdataset.py:
```diff
@@ -0,0 +1,695 @@
+from abc import ABC, abstractmethod
+
+
+class RawDataset(BaseDataset, ABC):
```
In the previous PR#161 I named this class `AbstractRawDataset`. Should we preserve that denomination?

jychoi-hpc replied:
Agree. I will rename.

allaffa commented on hydragnn/utils/rawdataset.py:
```diff
@@ -0,0 +1,695 @@
+                                self.dataset.append(data_object)
+            torch.distributed.barrier()
+
+            if self.data_format == "LSMS":
```
In PR#161 I had moved this `if` condition out of the AbstractRawDataset and made it part of the `LSMSDataset` class. Is it possible to do the same here?

jychoi-hpc replied:
Thank you for catching. I moved "charge_density_update_for_LSMS" to lsmsdataset but didn't delete here. I will clean up.

allaffa reviewed (changes requested):
First round of reviews.

jychoi-hpc commented on hydragnn/utils/rawdataset.py:
```diff
@@ -0,0 +1,695 @@
+            self.world_size = torch.distributed.get_world_size()
+            self.rank = torch.distributed.get_rank()
+
+        self.__load_raw_data(sampling=self.sampling)
```
Make sense. I will change.

jychoi-hpc commented on hydragnn/utils/rawdataset.py:
```diff
@@ -0,0 +1,695 @@
+
+        ## SerializedDataLoader
+        self.verbosity = config["Verbosity"]["level"]
+        self.node_feature_name = config["Dataset"]["node_features"]["name"]
+        self.node_feature_dim = config["Dataset"]["node_features"]["dim"]
+        self.node_feature_col = config["Dataset"]["node_features"]["column_index"]
+        self.graph_feature_name = config["Dataset"]["graph_features"]["name"]
+        self.graph_feature_dim = config["Dataset"]["graph_features"]["dim"]
+        self.graph_feature_col = config["Dataset"]["graph_features"]["column_index"]
```
Thank you for checking. I will remove.

jychoi-hpc commented on hydragnn/utils/rawdataset.py:
```diff
@@ -0,0 +1,695 @@
+        # In situations where someone already provides the .pkl filed with data
+        # the asserts from raw_dataset_loader are not performed
+        # Therefore, we need to re-check consistency
+        assert len(self.node_feature_name) == len(self.node_feature_dim)
+        assert len(self.node_feature_name) == len(self.node_feature_col)
+        assert len(self.graph_feature_name) == len(self.graph_feature_dim)
+        assert len(self.graph_feature_name) == len(self.graph_feature_col)
```
You are right. It's repeating. I will remove.

jychoi-hpc commented on hydragnn/utils/rawdataset.py:
```diff
@@ -0,0 +1,695 @@
+        feature_indices = [i for i in atom_features]
+        data.x = data.x[:, feature_indices]
+
+    def __load_serialized_data(self):
```
How about "build_edge"?

jychoi-hpc commented on hydragnn/utils/rawdataset.py:
```diff
@@ -0,0 +1,695 @@
+    return subsample
+
+
+class LSMSDataset(RawDataset):
```
Agree. Will make separate files.

jychoi-hpc commented on examples/lsms/lsms.py:
```diff
@@ -67,85 +74,87 @@ def info(*args, logtype="info", sep=" "):
+        trainset = SerializedDataset(basedir, datasetname, "trainset")
+        valset = SerializedDataset(basedir, datasetname, "valset")
+        testset = SerializedDataset(basedir, datasetname, "testset")
+        config["NeuralNetwork"]["Variables_of_interest"]
```
My mistake. Something left over after copy and paste. I will remove.

pzhanggit replied:
@jychoi-hpc  Just noticed this line is still there?

jychoi-hpc commented on hydragnn/utils/rawdataset.py:
```diff
@@ -0,0 +1,695 @@
+    return torch.from_numpy(np.divide(x1, x2, out=np.zeros_like(x1), where=x2 != 0))
+
+
+def nsplit(a, n):
```
Good suggestion. I agree. Will make it as a common function.

jychoi-hpc commented on hydragnn/utils/rawdataset.py:
```diff
@@ -0,0 +1,695 @@
+    def transform_input_to_data_object_base(self, filepath):
+        pass
+
+    def __charge_density_update_for_LSMS(self, data_object: Data):
```
Same as above. I already put "charge_density_update_for_LSMS" function to lsms class but didn't clean here. I will update.

jychoi-hpc commented:
The CI test error was caused because multiple processes overwrote "config.json" and left it as a corrupted file. I made a common function `save_config` and replaced throughout the codes.

I think I addressed all the previous comments. Please check again and leave further comments.

pzhanggit reviewed (commented):
Thanks for fixing the comments. It looks good to me in general. I added a few more comments. Most of them are minor, except for the one in `lsms.py`. I think the current implementation might cause trouble in `pickle` and `mpirun -n 2` mode.

jychoi-hpc commented on hydragnn/utils/model.py:
```diff
@@ -117,3 +119,16 @@ def print_model(model):
+    return torch.from_numpy(np.divide(x1, x2, out=np.zeros_like(x1), where=x2 != 0))
+
+
+def save_config(config, log_name, path="./logs/"):
```
Totally agree. Will move.

jychoi-hpc commented on examples/ising_model/train_ising.py:
```diff
@@ -201,41 +215,77 @@ def nsplit(a, n):
+        fname = os.path.join(os.path.dirname(__file__), "./dataset/%s.bp" % modelname)
+        trainset = AdiosDataset(fname, "trainset", comm, **opt)
+        valset = AdiosDataset(fname, "valset", comm, **opt)
+        testset = AdiosDataset(fname, "testset", comm, **opt)
```
Adios reads "minmax" automatically. It is same for SimplePickleDataset. I think the following line made a confusion:
```
minmax_node_feature = trainset.minmax_node_feature
```
It is not needed anymore. I will remove

jychoi-hpc commented:
@pzhanggit I addressed your second reviews. Please leave further comments.

pzhanggit commented:
> @pzhanggit I addressed your second reviews. Please leave further comments.

@jychoi-hpc  have you pushed any new changes? I could not see them from my side.

jychoi-hpc commented:
@pzhanggit Sorry. I just updated. Can you check again?

pzhanggit reviewed (approved):
Thanks for fixing my comments! It looks all good to me.

allaffa commented:
@pzhanggit @jychoi-hpc 
Merged :-)
